### ISSUE TITLE
[Custom Training] Dataset image size

### ISSUE THREAD

**Issue by matias-kovero** [ultralytics/yolov5#4401]
## ❔Question
**Hi**,
tried to find info on this on past issues, but can't really find anything - so I'll ask it here. 
Hopefully this will aid someone else if they are strugling with the same issue.

### Choosing image size.
My case, I would like to detect objects from cctv feed. This feed is native _1920x1080_. 
For speed(?) I could inference an section (point of intrest) of this feed, ex: an doorway etc.
This section could be ex: _500x500_ area of the original feed.

I have a bunch of legacy feed to create an dataset - this feed is in _1920x1080_.
What size images should I use in my dataset?
- 1920x1080
- 500x500
- Other?

I know I have the ability to select image size when training with `--img 'size'`

Should I just label my images in _1920x1080_ - then maybe scale them down to `1280x720 | 854x480 | 640x360 | 426x240`

> _Scale down as much as possible to speed up training_  

However, if I have small objects in the dataset, I can't scale down that much as it will hinder accuracy, am I right?
Also as my dataset images would be in 16:9 and not in 1:1, I should use `--rect`?

Finally, am I destroying my model with these image sizes, if ultimately I'm trying to inference on an 500x500 _(point of intrest)_ section of the 1920x1080 feed?

**Comment by glenn-jocher** [comment]
@matias-kovero 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by matias-kovero** [comment]
Hi @glenn-jocher, I have already read the Train Custom Data info. 
I know you are super busy, and these questions overflow this section.

Maybe it is because I'm not native english, but my earlier questions was aimed on this section that I don't fully grasp
> COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at --img 1280 you should also test and detect at `--img 1280`.

My dataset images are 1920x1080
-  I could benefit on making them initially smaller to start training the model quickly i.e `--img 1280`
- As my images are in 16:9 ration I should use `--rect` when training?


> Best inference results are obtained at the same `--img` as the training was run

So I should give `1280x720` images to inference for best results?

But to reduce unnecessary data given to inference I could provide an cropped section from the original feed `(1920x1080)` to the inference.

This cropped section would be in 1920x1080 resolution, but cropped to ex. 500x500 (1:1) or an smaller (16:9) section.
Will I hinder my inference results if I use an cropped image (Point of intrest) to inferece or should I just provide it the raw feed?

**Comment by glenn-jocher** [comment]
@matias-kovero as the guide states questions are pointless without a baseline result first using default settings.

**Comment by matias-kovero** [comment]
Those questions where more on the theory, as if _is my process fundamentally flawed?_

When using 16:9 images, should you use `--rect` on `train.py` _<- baseline needed for this?_

- **raw image**: `1920x1080`
- **cropped image**: `500x500`

Using **cropped image** from **raw image** for inference will give bad results if model trained with `--img 1280`?

**Comment by glenn-jocher** [comment]
I recommend following our published guidance

>  we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

**Comment by rsomani95** [comment]
@glenn-jocher FWIW I think it would be useful to know how to train with rectangular images without establishing a baseline first too. You may be using the same model for multiple tasks, and if other tasks require a specific image size, it'd be useful to know how to do that from the get go.

Intuitively, images that may be at odd aspect ratios like 2.39:1 make no sense to train at 640x640 because a tall box could get squished beyond comprehension. Given resource constraints, one may not have the time to try the recommended 640x first too.

The `train.py`'s argparser help message (and info from previous issues) tells me that you can pass `--rect` to train with rectangular images, but the `--img-size` argument only seems to accept a single number for train/test size. Is this the shorter side of the image? Glancing through the `dataset.py` code [here](https://github1s.com/ultralytics/yolov5/blob/HEAD/utils/datasets.py#L630-L649) left me a little confused on how this is handled.

**TLDR:**
Regardless of recommendation, can you force the training image size to be a specific height and width for all images?

Many thanks!

**Comment by Camilochiang** [comment]
@rsomani95 I hope that you find the answer to your question. If is not to late see [this comment](https://github.com/ultralytics/yolov5/issues/2009#issuecomment-766147324) for details.
